python/prima.py

```diff
@@ -0,0 +1,341 @@
+    x: list
+    success: bool
+    status: int
+    message: str
+    fun: float
+    nfev: int
+
+
+def bobyqa(f, n, x0, xl, xu, rhobeg=1.0, rhoend=1e-3, maxfun=1000, iprint=0):
+    """
+    BOBYQA solver.
+
+    Parameters
+    ----------
+    f : callable
+        objective function, returns a float
+    n : int
+        input dimension
+    x0 : sequence of float
+        starting point
+    xl, xu : sequence of float
+        x bounds
+    rhobeg, rhoend : float
+        trust region radius
+    maxfun : int
+        max function evaluation
+    iprint : int
+        verbosity level
+
+    Returns
+    -------
+    res : Result
+        Result class
+    """
+    assert len(x0) == n
+    assert len(xl) == n
+    assert len(xu) == n
+    cn = ct.c_int(n)
+    cx = (ct.c_double * n)(*x0)
+    cf = ct.pointer(ct.c_double())
+    cxl = (ct.c_double * n)(*xl)
+    cxu = (ct.c_double * n)(*xu)
+    cnf = ct.pointer(ct.c_int())
+    crhobeg = ct.c_double(rhobeg)
+    crhoend = ct.c_double(rhoend)
+    cmaxfun = ct.c_int(maxfun)
+    ciprint = ct.c_int(iprint)
+    cfun = _prima_pure_callback(f, n)
+    status = _handle.prima_bobyqa(cfun, cn, cx, cf, cxl, cxu,
+            cnf, crhobeg, crhoend, cmaxfun, ciprint)
+    x = [cx[i] for i in range(n)]
+    success = (status == 0)
+    cstatus = ct.c_int(status)
+    message = _handle.prima_get_rc_string(cstatus).decode()
+    fun = cf[0]
+    nfev = cnf[0]
+    return Result(x, success, status, message, fun, nfev)
+
+
+def cobyla(m_nlcon, f, n, x0, m_ineq=0, Aineq=[], bineq=[], m_eq=0, Aeq=[], beq=[], xl=None, xu=None, rhobeg=1.0, rhoend=1e-3, maxfun=1000, iprint=0):
+    """
+    COBYLA solver.
+
+    Parameters
+    ----------
+    m_nlcon : int
+        non-linear constraint dimension
+    f : callable
+        objective function, returns a tuple consisting of a float
+        and the constraint values as a sequence of float of size m_nlcon
+    n : int
+        input dimension
+    x0 : sequence of float
+        starting point
+    m_ineq : int
+        Number of inequality constraints
+    m_eq : int
+        Number of equality constraints
+    xl, xu : sequence of float
+        x bounds
+    rhobeg, rhoend : float
+        trust region radius
+    maxfun : int
+        max function evaluation
+    iprint : int
+        verbosity level
+
+    Returns
+    -------
+    res : Result
+        Result class
+    """
+    cm_nlcon = ct.c_int(m_nlcon)
+    cfun_con = _prima_pure_con_callback(f, n, m_nlcon)
+    assert len(x0) == n
+    cn = ct.c_int(n)
+    cx = (ct.c_double * n)(*x0)
+    cf = ct.pointer(ct.c_double())
+    ccstrv = ct.pointer(ct.c_double())
+    nl=[0.0] * m_nlcon
+    cnlconstr = (ct.c_double * m_nlcon)(*nl)
+    cm_ineq = ct.c_int(m_ineq)
+    cAineq = (ct.c_double * (m_ineq * n))(*Aineq)
+    cbineq = (ct.c_double * n)(*bineq)
+    cm_eq = ct.c_int(m_eq)
+    cAeq = (ct.c_double * (m_eq * n))(*Aeq)
+    cbeq = (ct.c_double * n)(*beq)
+    xl = [-1e100] * n if xl is None else xl
```

Comment:
Is there an interfacing issue with C if using `float('-inf')`?

Reply:
lets use sys.float_info.max